recipe/2024-02-terraform-providers.md

```diff
@@ -0,0 +1,690 @@
+
+
+#### User story 2
+As an operator, I would like to manage cloud resources in different geographical regions. To enable this, I need to set up multiple configurations for the same provider for different regions and use an alias to refer to individual provider configurations. 
+
+## Design
+### Design details
+
+Reviewing popular provider configurations (GCP, Oracle, Heroku, DigitalOcean, Docker etc.) a large number of provider configurations can be setup by handling a combination of key value pairs, key value pairs in nested objects, secrets and environment variables.
+
+### Key-value pairs
+Attributes or settings, represented as key-value pairs are often used to define configuration parameters in provider configurations.
+We will parse and save these values in the Terraform configuration file, main.tf in a working directory, which the system creates today. 
+
+### Alias
+
+Aliases in Terraform allow users to manage multiple instances of the same provider with different configurations for a single module. 
+Ref links: 
+- [Alias Multiple Provider Configurations](https://developer.hashicorp.com/terraform/language/providers/configuration#alias-multiple-provider-configurations)
+- [Passing Providers Explicitly](https://developer.hashicorp.com/terraform/language/modules/develop/providers#passing-providers-explicitly)
+
+For a Terraform recipe, aliases needed by the module are declared in the required_providers block, where each alias is a unique provider configuration.
+eg.
+
+```
+terraform {
+  required_providers {
+    mycloud = {
+      source  = "mycorp/mycloud"
+      version = "~> 1.0"
+      configuration_aliases = [ mycloud.west_region, mycloud.east_region ]
+    }
+  }
+}
+// The `required_providers` block for this module specifies that it will need provider configuration for aliases west_region and east_region to deploy successfully.
+```
+
+The bicep configuration for Terraform providers described in the environment resource is typically written to `provider` block in the main.tf.json file. These provider configurations are then passed to the recipe module. For provider configurations without aliases, they are be passed to the module automatically by Terraform.
+
+However, when the recipe module specifies one or more aliases in the `required_providers` block (and the module does not contain the provider configuration details), the aliased provider configuration from the `provider` block needs to be passed explicitly to the module.
+
+This is done by updating the module block in the main.tf.json file with provider aliases.
+
+#### Use Case 1: Matching aliases between provider configuration and `required_provider` block: sample generated main.tf.json file
+
+With current design, we are able to update the module configuration block with provider aliases when a match is found.
```

Comment:
This choice of design is not something we can revert back from if we decide to let user control the mappings entirely, so let's make sure we think about long term goals before deciding that this is an appropriate intermediate solution.